### PR TITLE
Manual Visibility

### PR DESCRIPTION
**Pull request by OakLoaf**
Added and implemented manual visibility.

Discussion is required regarding the ManualVisibility class, I feel like it is fine as it is, but just not sure whether the contents would be preferred to be moved into the enum.

This needs testing

### CODE REVIEW

File: api/src/main/java/de/oliver/fancyholograms/api/hologram/Hologram.java [hunk 1]
```diff
@@ -113,6 +114,10 @@ public final void showHologram(Player player) {
      */
     public final void forceShowHologram(Player player) {
         show(player);
+
+        if (this.getData().getVisibility().equals(Visibility.MANUAL)) {
+            Visibility.ManualVisibility.removeDistantViewer(this, player.getUniqueId());
```
Comment (OliverSchlueter) [inline api/src/main/java/de/oliver/fancyholograms/api/hologram/Hologram.java:119]: why remove here?

File: api/src/main/java/de/oliver/fancyholograms/api/data/property/Visibility.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        public static void addDistantViewer(Hologram hologram, UUID uuid) {
+            addDistantViewer(hologram, uuid);
+        }
+
+        public static void addDistantViewer(String hologramName, UUID uuid) {
+            distantViewers.put(hologramName, uuid);
+        }
+
+        public static void removeDistantViewer(Hologram hologram, UUID uuid) {
+            removeDistantViewer(hologram, uuid);
```
Comment (OliverSchlueter) [inline api/src/main/java/de/oliver/fancyholograms/api/data/property/Visibility.java:69]: shouldn't it be hologram.getName() or is the toString method overwritten to the name?

File: api/src/main/java/de/oliver/fancyholograms/api/data/property/Visibility.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    public static class ManualVisibility {
+        private static final HashMultimap<String, UUID> distantViewers = HashMultimap.create();
+
+        public static boolean canSee(Player player, Hologram hologram) {
+            return hologram.isViewer(player) || distantViewers.containsEntry(hologram.getName(), player.getUniqueId());
+        }
+
+        public static void addDistantViewer(Hologram hologram, UUID uuid) {
+            addDistantViewer(hologram, uuid);
```
Comment (OliverSchlueter) [inline api/src/main/java/de/oliver/fancyholograms/api/data/property/Visibility.java:61]: shouldn't it be hologram.getName(), so it goes into the addDistantViewer method with the string as parameter?

File: api/src/main/java/de/oliver/fancyholograms/api/data/HologramData.java [hunk 4]
```diff
@@ -101,8 +101,15 @@ public Visibility getVisibility() {
      * Set the type of visibility for the hologram.
      */
     public HologramData setVisibility(Visibility visibility) {
+        if (this.visibility != visibility) {
+            if (this.visibility.equals(Visibility.MANUAL)) {
```
Comment (OliverSchlueter) [inline api/src/main/java/de/oliver/fancyholograms/api/data/HologramData.java:105]: I guess this should be Objects.equals too then

Comment (OakLoaf) [reply api/src/main/java/de/oliver/fancyholograms/api/data/HologramData.java:105]: Visibility shouldn't be null so I think this is fine, I will add a NotNull annotation though

**Comment by OakLoaf** [comment]
All of those should be resolved now, will have to fix conflicts later though

**Comment by OakLoaf** [comment]
Done already lol

**Review by OliverSchlueter — APPROVED** [review]
LGTM

**Pull request merged**